c_common/front_end_common_lib/include/spinn_extra.h

```diff
@@ -0,0 +1,1106 @@
+/*
+ * Copyright (c) 2019-2020 The University of Manchester
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// ------------------------------------------------------------------------
+//
+// Extra definitions of things on SpiNNaker chips that aren't already mentioned
+// in spinnaker.h, or where the description is miserable.
+//
+// ------------------------------------------------------------------------
+
+#ifndef __SPINN_EXTRA_H__
+#define __SPINN_EXTRA_H__
+
+#include <spinnaker.h>
+#include <stdbool.h>
+#if defined(__GNUC__) && __GNUC__ < 6
+// This particular warning (included in -Wextra) is retarded wrong for client
+// code of this file. Only really a problem on Travis.
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+#endif // __GNUC__
+
+// Generates valid code if the named type is one word long, and invalid code
+// otherwise. It would be simpler if we used _Static_assert, but Eclipse seems
```

Comment:
The problem is that `static_assert()` only compiles correctly when put inside a function (don't know why; that'd seem to be something sensible to have outside a function as it is a compile-time constraint!), and this header file doesn't declare any functions. 

The replacement magic works and lets us ensure that things which are supposed to be word-sized (because they're describing the bit-layout of a register) are actually just that. Failing to have that assertion cost me over a day of debugging time…

Reply:
Correction, `_Static_assert` (note prefix) works fine outside functions from at least GCC 5.2.1 onwards (that's the earliest version I can easily check online), `static_assert` would but we can't include `assert.h` (because spinn_common interferes), and the Eclipse parser throws a wobbly about it and that's insanely distracting. But putting it in an otherwise-vacuous function works very well as an alternative, generating a helpful error message when the Bad Case happens. (In theory I could put all that stuff at the end of the header file… but I prefer keeping such things closer at hand.)

Reply:
BTW, I have tested the assertion code. I'm just not going to commit the test precisely because when it fires off something is truly broken. 😁